server/sonar-server/src/main/java/org/sonar/server/design/ws/DsmAction.java

```diff
@@ -0,0 +1,183 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.design.ws;
+
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Maps;
+import com.google.common.io.Resources;
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.api.utils.text.JsonWriter;
+import org.sonar.api.web.UserRole;
+import org.sonar.core.component.ComponentDto;
+import org.sonar.core.measure.db.MeasureDto;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.util.NonNullInputFunction;
+import org.sonar.core.util.NonNullInputPredicate;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.design.db.DsmDataEncoder;
+import org.sonar.server.design.db.DsmDb;
+import org.sonar.server.measure.ServerMetrics;
+import org.sonar.server.user.UserSession;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+public class DsmAction implements DependenciesAction {
+
+  private static final String PARAM_COMPONENT_UUID = "componentUuid";
+  private static final String PARAM_DISPLAY_NO_DEP = "displayComponentsWithoutDep";
+
+  private static Comparator<ComponentDto> COMPONENT_BY_NAME_COMPARATOR = new Comparator<ComponentDto>() {
+    @Override
+    public int compare(ComponentDto c1, ComponentDto c2) {
+      return c1.name().compareTo(c2.name());
+    }
+  };
+
+  private final DbClient dbClient;
+
+  public DsmAction(DbClient dbClient) {
+    this.dbClient = dbClient;
+  }
+
+  @Override
+  public void define(WebService.NewController controller) {
+    WebService.NewAction action = controller.createAction("dsm")
+      .setDescription("Display the DSM")
+      .setSince("5.2")
+      .setInternal(true)
+      .setHandler(this)
+      .setResponseExample(Resources.getResource(getClass(), "example-dsm.json"));
+
+    // TODO add component key param
+    action.createParam(PARAM_COMPONENT_UUID)
+      .setDescription("Component uuid (only sub project or directory are allowed for the moment)")
+      .setRequired(true)
+      .setExampleValue("2312cd03-b514-4acc-94f4-5c5e8e0062b2");
+
+    action.createParam(PARAM_DISPLAY_NO_DEP)
+      .setDescription("Display component with no dependencies or not")
```

Comment:
Because displaying them makes the DSM unreadable when there's a lof of directories / files. Currently in the UI there's a checkbox to display them.